photon-lib/src/main/java/org/photonvision/simulation/VisionSystemSim.java

```diff
@@ -0,0 +1,379 @@
+     * @return If the cameraSim was valid and transform was adjusted
+     */
+    public boolean adjustCamera(PhotonCameraSim cameraSim, Transform3d robotToCamera) {
+        var trfBuffer = camTrfMap.get(cameraSim);
+        if (trfBuffer == null) return false;
+        trfBuffer.addSample(Timer.getFPGATimestamp(), new Pose3d().plus(robotToCamera));
+        return true;
+    }
+
+    /** Reset the previous transforms for all cameras to their current transform. */
+    public void resetCameraTransforms() {
+        for (var cam : camTrfMap.keySet()) resetCameraTransforms(cam);
+    }
+
+    /**
+     * Reset the transform history for this camera to just the current transform.
+     *
+     * @return If the cameraSim was valid and transforms were reset
+     */
+    public boolean resetCameraTransforms(PhotonCameraSim cameraSim) {
+        double now = Timer.getFPGATimestamp();
+        var trfBuffer = camTrfMap.get(cameraSim);
+        if (trfBuffer == null) return false;
+        trfBuffer.clear();
+        var lastTrf = new Transform3d(new Pose3d(), trfBuffer.getSample(now).orElse(new Pose3d()));
+        adjustCamera(cameraSim, lastTrf);
+        return true;
+    }
+
+    public Set<VisionTargetSim> getVisionTargets() {
+        var all = new HashSet<VisionTargetSim>();
+        for (var entry : targetSets.entrySet()) {
+            all.addAll(entry.getValue());
+        }
+        return all;
+    }
+
+    public Set<VisionTargetSim> getVisionTargets(String type) {
+        return targetSets.get(type);
+    }
+
+    /**
+     * Adds targets on the field which your vision system is designed to detect. The {@link
+     * PhotonCamera}s simulated from this system will report the location of the camera relative to
+     * the subset of these targets which are visible from the given camera position.
+     *
+     * <p>By default these are added under the type "targets".
+     *
+     * @param targets Targets to add to the simulated field
+     */
+    public void addVisionTargets(VisionTargetSim... targets) {
+        addVisionTargets("targets", targets);
+    }
+
+    /**
+     * Adds targets on the field which your vision system is designed to detect. The {@link
+     * PhotonCamera}s simulated from this system will report the location of the camera relative to
+     * the subset of these targets which are visible from the given camera position.
+     *
+     * <p>The AprilTags from this layout will be added as vision targets under the type "apriltags".
+     * The poses added preserve the tag layout's current alliance origin.
+     *
+     * @param tagLayout The field tag layout to get Apriltag poses and IDs from
+     */
+    public void addVisionTargets(AprilTagFieldLayout tagLayout) {
+        for (AprilTag tag : tagLayout.getTags()) {
+            addVisionTargets(
+                    "apriltags",
+                    new VisionTargetSim(
+                            tagLayout.getTagPose(tag.ID).get(), // preserve alliance rotation
```

Comment:
Check if tag is present, getTagPose returns an optional

Reply:
We are iterating through the tags that the tag layout reports through `getTags()` and then redundantly getting their pose from the layout again (for handling different alliance origins), so this should not be empty.